cpp/tests/experimental/spatial/linestring_distance_test.cu

```diff
@@ -0,0 +1,153 @@
+  using CartVec = std::vector<cartesian_2d<T>>;
+
+  auto cart2ds = rmm::device_vector<cartesian_2d<T>>{
+    CartVec({{0.0f, 0.0f}, {1.0f, 0.0f}, {2.0f, 0.0f}, {3.0f, 0.0f}, {4.0f, 0.0f}})};
+  auto offset = rmm::device_vector<int32_t>{std::vector<int32_t>{0}};
+
+  auto a_begin = cart2ds.begin();
+  auto a_end   = cart2ds.begin() + 3;
+  auto b_begin = cart2ds.begin() + 1;
+  auto b_end   = cart2ds.end();
+
+  auto distance = rmm::device_vector<T>{std::vector<T>{0.0}};
+  auto expected = rmm::device_vector<T>{std::vector<T>{0.0}};
+
+  pairwise_linestring_distance(
+    offset.begin(), offset.end(), a_begin, a_end, offset.begin(), b_begin, b_end, distance.begin());
+
+  EXPECT_EQ(distance, expected);
+}
+
+TYPED_TEST(PairwiseLinestringDistanceTest, FromTransformIterator)
+{
+  using T       = TypeParam;
+  using CartVec = std::vector<cartesian_2d<T>>;
+
+  auto a_cart2d_x = rmm::device_vector<T>{std::vector<T>{0.0, 1.0, 2.0, 3.0, 4.0}};
+  auto a_cart2d_y = rmm::device_vector<T>(5, 0.0);
+
+  auto a_begin = make_cartesian_2d_iterator(a_cart2d_x.begin(), a_cart2d_y.begin());
+  auto a_end   = a_begin + a_cart2d_x.size();
+
+  auto b_cart2d_x = rmm::device_vector<T>{std::vector<T>{0.0, 1.0, 2.0, 3.0, 4.0}};
+  auto b_cart2d_y = rmm::device_vector<T>(5, 1.0);
+
+  auto b_begin = make_cartesian_2d_iterator(b_cart2d_x.begin(), b_cart2d_y.begin());
+  auto b_end   = b_begin + b_cart2d_x.size();
+
+  auto offset = rmm::device_vector<int32_t>{std::vector<int32_t>{0}};
+
+  auto distance = rmm::device_vector<T>{std::vector<T>{0.0}};
+  auto expected = rmm::device_vector<T>{std::vector<T>{1.0}};
+
+  pairwise_linestring_distance(
+    offset.begin(), offset.end(), a_begin, a_end, offset.begin(), b_begin, b_end, distance.begin());
+
+  EXPECT_EQ(distance, expected);
+}
+
+TYPED_TEST(PairwiseLinestringDistanceTest, FromMixedIterator)
+{
+  using T       = TypeParam;
+  using CartVec = std::vector<cartesian_2d<T>>;
+
+  auto a_cart2d = rmm::device_vector<cartesian_2d<T>>{
+    CartVec({{0.0f, 0.0f}, {1.0f, 0.0f}, {2.0f, 0.0f}, {3.0f, 0.0f}, {4.0f, 0.0f}})};
+
+  auto b_cart2d_x = rmm::device_vector<T>{std::vector<T>{0.0, 1.0, 2.0, 3.0, 4.0}};
+  auto b_cart2d_y = rmm::device_vector<T>(5, 1.0);
+
+  auto b_begin = make_cartesian_2d_iterator(b_cart2d_x.begin(), b_cart2d_y.begin());
+  auto b_end   = b_begin + b_cart2d_x.size();
+
+  auto offset = rmm::device_vector<int32_t>{std::vector<int32_t>{0}};
+
+  auto distance = rmm::device_vector<T>{std::vector<T>{0.0}};
+  auto expected = rmm::device_vector<T>{std::vector<T>{1.0}};
+
+  pairwise_linestring_distance(offset.begin(),
+                               offset.end(),
+                               a_cart2d.begin(),
+                               a_cart2d.end(),
+                               offset.begin(),
+                               b_begin,
+                               b_end,
+                               distance.begin());
+
+  EXPECT_EQ(distance, expected);
+}
+
```

Comment:
Anytime we have a custom CUDA kernel or significant custom CUDA logic (like a non-trivial operator passed to a Thrust algorithm), I think we should have at least one test that causes the kernel to be invoked with more threads than there are in a single warp. This is especially true when the kernel involves cooperation between threads, as you do in this algorithm's kernel using atomic operations. So I suggest making a test that generates at least ~1000 data points.